core/embed/trezorhal/flash.h

```diff
@@ -24,71 +24,41 @@
-#define FLASH_SECTOR_COUNT 24
-#elif defined STM32F405xx
-#define FLASH_SECTOR_COUNT 12
-#else
-#error Unsupported MCU
-#endif
+#define FLASH_OTP_NUM_BLOCKS 16
+#define FLASH_OTP_BLOCK_SIZE 32
 
-#define FLASH_SECTOR_BOARDLOADER_START 0
-//                                           1
-#define FLASH_SECTOR_BOARDLOADER_END 2
-
-//                                           3
-
-#if defined TREZOR_MODEL_T || defined TREZOR_MODEL_R || \
-    defined TREZOR_MODEL_DISC1
-#define FLASH_SECTOR_STORAGE_1 4
-#define FLASH_SECTOR_STORAGE_2 16
-#elif defined TREZOR_MODEL_1
-#define FLASH_SECTOR_STORAGE_1 2
-#define FLASH_SECTOR_STORAGE_2 3
-#else
-#error Unknown Trezor model
+#ifndef STM32U5A9xx
+#define FLASH_BYTE_ACCESS 1
 #endif
 
-#define FLASH_SECTOR_BOOTLOADER 5
-
-#define FLASH_SECTOR_FIRMWARE_START 6
-//                                           7
-//                                           8
-//                                           9
-//                                          10
-#define FLASH_SECTOR_FIRMWARE_END 11
-
-#define FLASH_SECTOR_UNUSED_START 12
-//                                          13
-//                                          14
-#define FLASH_SECTOR_UNUSED_END 15
-
-#define FLASH_SECTOR_FIRMWARE_EXTRA_START 17
-//                                          18
-//                                          19
-//                                          20
-//                                          21
-//                                          22
-#define FLASH_SECTOR_FIRMWARE_EXTRA_END 23
-
-#define BOOTLOADER_SECTORS_COUNT (1)
-#define STORAGE_SECTORS_COUNT (2)
-#define FIRMWARE_SECTORS_COUNT (6 + 7)
-
-extern const uint8_t STORAGE_SECTORS[STORAGE_SECTORS_COUNT];
-extern const uint8_t FIRMWARE_SECTORS[FIRMWARE_SECTORS_COUNT];
-
-// note: FLASH_SR_RDERR is STM32F42xxx and STM32F43xxx specific (STM32F427)
-// (reference RM0090 section 3.7.5)
-#if !defined STM32F427xx && !defined STM32F429xx
-#define FLASH_SR_RDERR 0
-#endif
+/**
+ * Flash driver interface is designed to abstract away differences between
+ * various MCUs used in Trezor devices.
+ *
+ * Generally, flash memory is divided into sectors. On different MCUs, sectors
+ * may have different sizes, and therefore, different number of sectors are used
+ * for a given purpose. For example, on STM32F4, the sectors are relatively
+ * large so we use single sector for Storage. On STM32U5, the sectors are
+ * smaller, so we use multiple sectors for the Storage. Storage implementation
+ * should not care about this, and should use flash_area_t interface to access
+ * the flash memory.
+ *
+ * flash_area_t represents a location in flash memory. It may be contiguous, or
+ * it may be composed of multiple non-contiguous subareas.
+ *
+ * flash_subarea_t represents a contiguous area in flash memory, specified by
+ * first_sector and num_sectors.
+ */
+
+typedef struct {
+  uint16_t first_sector;
+  uint16_t num_sectors;
+} flash_subarea_t;
```

Comment:
I can see that `flash_subarea_t` and `flash_area_t` are defined in three places (here, `legacy/flash.h` and in `storage/tests/c/flash.h`). Also the functions that are further down below seem to be declared three times. Is there a good reason for that? I'd expect there to be something like `storage/flash_common.h` which would declare these.

Reply:
well this was the case all along, this PR adds the new functions/definitions in same manner. i agree that there is a lot of repetition, but not sure what would be the best place for the common file - maybe trezor-common? (don't really like putting the common file into storage as the definition of flash interface is not really what would you expect from that)